clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -6598,48 +6683,88 @@ void Sema::applyPtrCountedByEndedByAttr(Decl *D, unsigned Level,
-        AttachDependerDeclsAttr(VD, BDTy, EffectiveLevel);
+      if (!diagnoseCountDependentDecls(*this, Info.VD, BDTy,
+                                       Info.EffectiveLevel, Info.IsFPtr))
+        AttachDependerDeclsAttr(Info.VD, BDTy, Info.EffectiveLevel);
     } else if (const auto *BDTy =
                    dyn_cast<DynamicRangePointerType>(ConstructedType)) {
-      diagnoseRangeDependentDecls(*this, VD, BDTy, EffectiveLevel, IsFPtr);
+      diagnoseRangeDependentDecls(*this, Info.VD, BDTy, Info.EffectiveLevel,
+                                  Info.IsFPtr);
     } else {
       llvm_unreachable("Unexpected bounds attributed type");
     }
   }
 
-  if (TND) {
+  if (Info.TND) {
     // We need to create a TypeSourceInfo, as TypedefNameDecl is not a ValueDecl
     // and thus doesn't have setType() method.
-    SourceLocation Loc = TND->getTypeSourceInfo()->getTypeLoc().getBeginLoc();
-    TypeSourceInfo *Info = Context.getTrivialTypeSourceInfo(NewDeclTy, Loc);
-    TND->setTypeSourceInfo(Info);
+    SourceLocation Loc =
+        Info.TND->getTypeSourceInfo()->getTypeLoc().getBeginLoc();
+    TypeSourceInfo *TSI = Context.getTrivialTypeSourceInfo(NewDeclTy, Loc);
+    Info.TND->setTypeSourceInfo(TSI);
   } else {
-    VD->setType(NewDeclTy);
+    Info.VD->setType(NewDeclTy);
     // Reconstruct implicit cast for initializer after variable type change.
-    if (Var && Var->hasInit()) {
-      Expr *Init = Var->getInit();
+    if (Info.Var && Info.Var->hasInit()) {
+      Expr *Init = Info.Var->getInit();
       ExprResult Res = removeUnusedOVEs(*this, Init, Init->IgnoreImpCasts());
       if (Res.isInvalid())
         return;
-      AddInitializerToDecl(Var, Res.get(), Var->isDirectInit());
+      AddInitializerToDecl(Info.Var, Res.get(), Info.Var->isDirectInit());
     }
   }
 }
 
+static void attachLateInstantiatedCountedByEndedByAttr(Sema &S, Decl *D,
+                                                       const ParsedAttr &AL,
+                                                       int Level) {
+#define ADD_ATTR(Kind)                                                         \
+  case ParsedAttr::AT_##Kind:                                                  \
+    D->addAttr(::new (S.Context)                                               \
+                   Kind##Attr(S.Context, AL, AL.getArgAsExpr(0), Level));      \
+    break;
+
+  switch (AL.getKind()) {
+    ADD_ATTR(SizedBy)
+    ADD_ATTR(SizedByOrNull)
+    ADD_ATTR(CountedBy)
+    ADD_ATTR(CountedByOrNull)
+    ADD_ATTR(PtrEndedBy)
+  default:
+    llvm_unreachable("Invalid dynamic bound attribute");
+  }
+#undef ADD_ATTR
+}
+
 static void handlePtrCountedByEndedByAttr(Sema &S, Decl *D,
                                           const ParsedAttr &AL) {
   unsigned Level;
   if (!S.checkUInt32Argument(AL, AL.getArgAsExpr(1), Level))
     return;
+
+  if (D->getDescribedTemplate() || S.CurContext->isDependentContext()) {
```

Comment:
Same question here, wondering if this is sufficient for non-templated entities nested in templates.

Reply:
Also replaced with a call to `isTemplated`